src/record.rs

```diff
@@ -1,47 +1,227 @@
+///
+/// It is guaranteed to be well-formed, but may not be valid according to the specification.
 #[derive(Clone, Debug, PartialEq)]
-pub struct Record {
-    /// The WARC standard version this record conforms to.
+pub struct RawRecord {
+    /// The WARC standard version this record reports conformance to.
     pub version: String,
-    /// A set of all headers that are part of this record.
-    pub headers: HashMap<crate::header::WarcHeader, Vec<u8>>,
+    /// All headers that are part of this record.
+    pub headers: HashMap<WarcHeader, Vec<u8>>,
     /// The data body of this record.
     pub body: Vec<u8>,
 }
 
+/// A builder for WARC records from data.
+pub struct RecordBuilder {
+    version: String,
+    headers: HashMap<WarcHeader, Vec<u8>>,
+    body: Vec<u8>,
+    record_date: Option<DateTime<Utc>>,
+    record_id: Option<String>,
+    record_type: Option<String>,
+}
+
+/// A single WARC record.
+///
+/// It is guaranteed to be valid according to the specification it conforms to, except:
+/// * The validity of the WARC-Record-ID header is not checked
+/// * Date information not in the UTC timezone will be silently converted to UTC
+///
+/// This record can be constructed by a `RecordBuilder` or by a fallable cast from a `RawRecord`.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Record {
+    // NB: does not contain the fields stored in the struct
+    raw: RawRecord,
+
+    record_date: DateTime<Utc>,
+    record_id: String,
+    record_type: String,
+}
+
 impl Record {
-    /// Return a new UUID as a string value for the WARC-Record-ID header.
-    pub fn make_uuid() -> String {
-        format!("<{}>", Uuid::new_v4().to_urn())
+    /// Create a new empty record with default values.
+    ///
+    /// Using a `RecordBuilder` is more efficient when creating records from known data.
+    ///
+    /// A default record contains an empty body, and the following fields:
+    /// * WARC-Record-ID: generated by `generate_record_id()`
+    /// * WARC-Date: the current moment in time
+    /// * WARC-Type: resource
+    /// * WARC-Content-Length: 0
+    pub fn new() -> Record {
+        Record::default()
+    }
+
+    /// Transform this record into a raw record containing the same data.
+    pub fn into_raw(self) -> RawRecord {
+        let Record { mut raw, record_date, record_id, record_type } = self;
+        let insert1 = raw.headers.insert(WarcHeader::ContentLength,
+                                         format!("{}", raw.body.len()).into());
+        let insert2 = raw.headers.insert(WarcHeader::WarcType, record_type.into());
+        let insert3 = raw.headers.insert(WarcHeader::RecordID, record_id.into());
+        let insert4 = raw.headers.insert(WarcHeader::Date, record_date.to_rfc3339().into());
+
+        debug_assert!(insert1.is_none() && insert2.is_none() &&
+                      insert3.is_none() && insert4.is_none(),
+                      "invariant violation: raw struct contains externally stored fields");
```

Comment:
I don't fully follow this `debug_assert`.

> If the map did not have this key present, `None` is returned.
https://doc.rust-lang.org/beta/std/collections/struct.HashMap.html#method.insert

So if this assertion is ever true that means the raw record didn't have all 4 headers? What does that signal to us?

Reply:
That is correct.

This assert is checking the invariant noted for the struct's internal `raw` field -- which I fixed a mistake in:
```
// NB: does not contain the headers stored in the struct
```

In other words, the struct is expected to keep an _invalid_ raw record, where all of the useful headers (some mandatory) are removed and kept to the struct's other fields instead. When the struct is converted into a raw record, this conversion must be undone by re-inserting those headers.

That conversion is the critical time to check the invariant: did any of the insertions of headers stored in the _replace instead of inserting anew_? Because if so, then we have duplicate, possibly differing data, and that's bad.

Since I expect it to be rare, and mostly useful during development, I made it a debug assert to avoid slowing down release builds.

Reply:
Yep that makes sense.